fairscale/experimental/wgit/signal_sparsity.py

```diff
@@ -4,10 +4,86 @@
 # LICENSE file in the root directory of this source tree.
 
 from enum import Enum
+from typing import Optional
 
+import torch
 from torch import Tensor
 
 
+# Helper Functions
+def _get_k_for_topk(topk_percent: Optional[float], top_k_element: Optional[int], top_k_total_size: int) -> int:
+    """Converts the top_k_percent to top_k_element when top_k_percent is provided
+    as the criterion for top-k calculation. When, top_k_element is used as the criterion,
+    simply returns the value for k. Also, ensures k is never 0 to avoid all sparse tensors.
+    """
+    if top_k_element is None:
+        top_k_element = int(top_k_total_size * topk_percent / 100.0)
+
+    # ensure we never have 100% sparsity in tensor and always have 1 surviving element!
+    top_k_element = max(1, top_k_element)
+    return top_k_element
+
+
+def _scatter_topk_to_sparse_tensor(
+    top_k_tensor: Tensor, to_be_sparsify_tensor: Tensor, k: int, dim: Optional[int]
+) -> Tensor:
+    """Scatter the topk values of the to_be_sparsify_tensor to a zero tensor of the same shape
+    at the top-k indices of the top_k_tensor. This function allows top-k computation with a
+    derived tensor from to_be_sparsify_tensor.
+
+    Args:
+        top_k_tensor (Tensor)
+            The source tensor whose top-k "indices" are taken and used to extract
+            the corresponding "values" from the to_be_sparsify_tensor.
+        to_be_sparsify_tensor (Tensor)
+            The tensor whose values are gathered according to the top-k indices
+            of the top_k_tensor, and a zero tensor of same shape is populated with these
+            values at those indices and creates the sparse_tensor tensor.
+        k (int)
+            the value of k for top-k
+        dim (int, optional)
+            dimension for top-k
+
+    Returns:
+        Returns a sparse_tensor with the same shape as the top_k_tensor and to_be_sparsify_tensor,
+        and populated with the values of the to_be_sparsify_tensor at the indices corresponding
+        to the top-k indices of the source tensor.
+    """
+    assert (
+        top_k_tensor.shape == to_be_sparsify_tensor.shape
+    ), "top_k_tensor and to_be_sparsify_tensor have different shapes!"
+
+    sparse_tensor = torch.zeros_like(to_be_sparsify_tensor)
+    if dim is None:
+        _, idx = top_k_tensor.flatten().topk(k)
+        sparse_tensor.flatten()[idx] = to_be_sparsify_tensor.flatten()[idx]
```

Comment:
don't we need to reshape the result back? That a bug here? Not tested?

I suggest the following structure:

```
shape = top_k_tensor.shape
if dim is None and len(shape)>1:
  orig_shape = ...
  sparse_tensor = sparse_tensor.reshape([-1]
  top_k_tensor = top_k_tensor.reshape...
  to_be ... = to_be...reshape...
  dim = -1

# then do gather-scatter uniformly

if orig_shape is not None:
  sparse_tensor = sparse_tensor.reshape(orig_shape)

```

Reply:
I don't think this is a bug, as Tensor.flatten doesn't mutate the tensor itself, but does in fact perform the value assignment to the indices. Can be replicated by this short snippet:
```
t = torch.ones(5,5)
print(t.shape)
t.flatten()[1] = 7
print(t)
```
I can however make this more explicit according to your suggestion. There is tests covering some aspects of this, but that relies on this strategy as well, so I will add more explicit tests for this. Thank you for pointing the test out.

Reply:
I see. You are right. I missed the fact that left-hand-side expression is just a temp tensor with the same backing.

I think explicit is better here and reshape() doesn't cost much since it should only change the tensor metadata a bit. Any code that is more obviously correct is better.